optuna/samplers/_nsgaiii.py

```diff
@@ -0,0 +1,637 @@
+        axis=2,
+    )
+    extreme_points = objective_matrix[np.argmin(asf_value, axis=1), :]
+
+    # Normalize objective_matrix with extreme points.
+    # Note that extreme_points can be degenerate, but no proper operation is remarked in the
+    # paper. Therefore, the maximum value of population in each axis is used in such cases.
+    if np.all(np.isfinite(extreme_points)) and np.linalg.matrix_rank(extreme_points) == len(
+        extreme_points
+    ):
+        intercepts_inv = np.linalg.solve(extreme_points, np.ones(n_objectives))
+    else:
+        intercepts_inv = 1 / np.max(objective_matrix, axis=0)
+    objective_matrix *= np.where(np.isfinite(intercepts_inv), intercepts_inv, 1)
+
+    return objective_matrix
+
+
+def _associate_individuals_with_reference_points(
+    objective_matrix: np.ndarray, reference_points: np.ndarray
+) -> tuple[np.ndarray, np.ndarray]:
+    """Associates each objective value to the closest reference point
+
+    Associate each normalized objective value to the closest reference point. The distance is
+    calculated by Euclidean norm.
+
+    Args:
+        objective_matrix:
+            A 2 dimension ``numpy.ndarray`` with columns of objective dimension and rows of
+            generation size. Each row is the normalized objective value of the corresponding
+            individual.
+
+    Returns:
+        closest_reference_points:
+            A ``numpy.ndarray`` with rows of generation size. Each row is the index of
+            the closest reference point to the corresponding individual.
+        distance_reference_points:
+            A ``numpy.ndarray`` with rows of generation size. Each row is the distance from
+            the corresponding individual to the closest reference point.
+    """
+    # TODO(Shinichi) Implement faster assignment for the default reference points because it does
+    # not seem necessary to calculate distance from all reference points.
+
+    # TODO(Shinichi) Normalize reference_points in constructor to remove reference_point_norms.
+    # In addition, the minimum distance from each reference point can be replaced with maximum
+    # inner product between the given individual and each normalized reference points.
+
+    # distance_from_reference_lines is a ndarray of shape (n, p), where n is the size of the
+    # population and p is the number of reference points. Its (i,j) entry keeps distance between
+    # the i-th individual values and the j-th reference line.
+    reference_point_norm_squared = np.linalg.norm(reference_points, axis=1) ** 2
+    perpendicular_vectors_to_reference_lines = np.einsum(
+        "ni,pi,p,pm->npm",
+        objective_matrix,
+        reference_points,
+        1 / reference_point_norm_squared,
+        reference_points,
+    )
+    distance_from_reference_lines = np.linalg.norm(
+        objective_matrix[:, np.newaxis, :] - perpendicular_vectors_to_reference_lines,
+        axis=2,
+    )
```

Comment:
ditto. This seems to correspond to L5 to 7 of Algorithm 3 in https://www.egr.msu.edu/~kdeb/papers/k2012009.pdf. We'd like to improve the readability in a follow-up PR.